### PR TITLE
[DPE-4138] Fix failures raised by integration tests on CI

### PR DESCRIPTION
**Pull request by Mehdi-Bendriss**
## Issue
This PR addresses [DPE-4138](https://warthogs.atlassian.net/browse/DPE-4138), namely, this PR:
- reworks the health API
- reworks applying health on various parts of the charm
- fixes the replication of each "built-in" index
- reworks "peer-rel" trigger logic
- removes the roles re-computing from the charm / tests
- fixes various bug
- fixes integration tests
- fixes client-integration test with failing secrets / requests 
- updates libs 

[DPE-4138]: https://warthogs.atlassian.net/browse/DPE-4138?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/charms/opensearch/v0/helper_charm.py [hunk 1]
```diff
@@ -115,3 +118,22 @@ def relation_departure_reason(charm: CharmBase, relation_name: str) -> RelDepart
         return RelDepartureReason.SCALE_DOWN
 
     return RelDepartureReason.REL_BROKEN
+
+
+def trigger_peer_rel_changed(
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/helper_charm.py:123]: This is a potential source of noise, why is this needed?

File: lib/charms/opensearch/v0/opensearch_health.py [hunk 2]
```diff
@@ -58,42 +61,87 @@ def apply(
...
+        # is being relocated. Examples are: simple stop, unit removal, upgrade
+        raise OpenSearchHAError("Shards haven't completed relocating.")
+
+    def _apply_for_app(self, status: str) -> None:
         """Cluster wide / app status."""
         if not self._charm.unit.is_leader():
-            # this is needed in case the leader is in an error state and doesn't
-            # report the status itself
-            self._charm.peers_data.put(Scope.UNIT, "health", status)
+            trigger_peer_rel_changed(self._charm, on_other_units=True)
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/opensearch_health.py:144]: If we will apply "something" in the peer relation, why not applying the `status`?

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py [hunk 3]
```diff
@@ -88,6 +89,11 @@ def run(self) -> None:
                 Scope.APP, "deployment-description", deployment_desc.to_dict()
             )
 
+        if deployment_desc.start == StartMode.WITH_GENERATED_ROLES:
+            # trigger roles change on the leader, other units will have their peer-rel-changed
+            # event triggered
+            trigger_peer_rel_changed(self._charm, on_other_units=False, on_current_unit=True)
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/opensearch_peer_clusters.py:95]: There is already a "put" above. Is that not enough?

**Review by phvalguima — APPROVED** [review]

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 4]
```diff
@@ -603,40 +607,49 @@ def _on_update_status(self, event: UpdateStatusEvent):
...
-            if self.plugin_manager.run():
+
+            if self.plugin_manager.run() and not restart_requested:
                 if self.upgrade_in_progress:
                     logger.warning(
                         "Changing config during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
                     )
                     event.defer()
                     return
+
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:652]: We should move any "plugin_manager / defer" *before* the line 629:
```
self.opensearch_peer_cm.run()
```
As that `run()` will make a call to `trigger_peer_rel_changed`

Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:652]: Please, also a very big comment banner informing: "no defer after this line", with an explanation on *why* we need that precaution.

File: lib/charms/opensearch/v0/helper_charm.py (hunk 1 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/helper_charm.py:123]: It's needed for 2 things: 
1. "roles changes" when manually requested by user (config change) - this is invoked, and triggers a peer rel changed event, where the whole topology is again recomputed and units restart
2. `health.apply(app=True)`, when called from a non leader unit - this unit computes the health but since it cannot write on the app status, it will trigger a peer rel changed event to be caught by the leader who will then compute the status and display it on the app level

File: lib/charms/opensearch/v0/opensearch_health.py [hunk 5]
```diff
@@ -58,42 +61,87 @@ def apply(
...
-            if not compute_health:
-                return HealthColors.IGNORE
+        status = self.get(
+            wait_for_green_first=wait_for_green_first,
+            use_localhost=use_localhost,
+        )
+        logger.info(f"Current health of cluster: {status}")
+
+        if app:
+            self._apply_for_app(status)
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/opensearch_health.py:71]: So, this OpenSearchHealth.apply, if calling apply, there is a potential to call the "trigger defer" method. We must make sure if: (1) we need this "trigger defer" in "_apply_for_app"; or (2) if we do, then we need to ensure there is no risk for any OpenSearchHealth.apply() to be called before a defer

File: lib/charms/opensearch/v0/opensearch_health.py (hunk 2 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_health.py:144]: Actually, this is never going to be called:
`_apply_for_app` only runs for app-level, this is supposed to never run at app-level (not leader).

Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_health.py:144]: Because the status is a variable that changes with time, setting/getting it from the peer rel data may lead to outdated status reporting (we used to do it like this on scale-down, it didn't work well)

**Review by phvalguima — CHANGES_REQUESTED** [review]
My main concern is the trigger mechanism. There is a potential of getting trapped between deferrals and continuously changing the peer relation data.

File: lib/charms/opensearch/v0/opensearch_health.py (hunk 2 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_health.py:144]: @Mehdi-Bendriss not sure if you saw my 2nd comment here:

> Actually, this is never going to be called:
_apply_for_app only runs for app-level, this is supposed to never run at app-level (not leader).

(It was 1 min apart from yours)

In a nutshell, if I understood correctly that should never be called.

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py (hunk 3 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py:95]: it would be but only if `deployment_desc.state.value == State.ACTIVE` - otherwise not

File: lib/charms/opensearch/v0/helper_charm.py (hunk 1 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/helper_charm.py:123]: If we are so coupled, why not running it all under the same config-changed?

Comment (phvalguima) [reply lib/charms/opensearch/v0/helper_charm.py:123]: If these events are so entangled together, why not running it all under the same config-changed?

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 4 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:652]: Not sure I understand what you mean exactly here.

> As that run() will make a call to trigger_peer_rel_changed

No - this will only run if there was a change to the 2 config options of the app:
- init-hold
- roles 
You can see this on the logic of `self.opensearch_peer_cm.run()` - it only concerns `exisiting clusters` which we're updating

File: lib/charms/opensearch/v0/opensearch_health.py [hunk 6]
```diff
@@ -58,42 +61,87 @@ def apply(
         app: bool = True,
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/opensearch_health.py:61]: @Mehdi-Bendriss we should default this to False, as `self._apply_for_app()` changes the peer relation directly

File: lib/charms/opensearch/v0/opensearch_health.py (hunk 5 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_health.py:71]: This will only happen in the case where, a NON-leader unit wants to apply the health to application status (which it cannot, so it emits this event so that the leader is notified that it must compute it). 

and this happens in exactly 2 cases:
- `_on_opensearch_data_storage_detaching`
- `_post_start_init`

File: lib/charms/opensearch/v0/opensearch_health.py (hunk 6 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_health.py:61]: Same answer as https://github.com/canonical/opensearch-operator/pull/279#discussion_r1589336622

File: lib/charms/opensearch/v0/helper_charm.py (hunk 1 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/helper_charm.py:123]: because this coming from when we were doing automated roles-rebalancing, which we're no longer doing sincea couple of weeks. This is something we're planning to address in the next cycle

**Review by phvalguima — APPROVED** [review]
The PR fixes most of the tests. The remaining will be investigated on `large-deployments-tests` PR. 

There is one concern I have we must keep a close watch: `trigger_peer_rel_changed` has the potential of endless calls to `opensearch-peers-changed` hook. Specially with `_post_start_init`.
